Docker image component

pmuens opened the pull request:
Implements a Docker image component which makes it possible to build and publish Docker images.

The following config can be used to test the `build` / `deploy` and `remove` commands:

```yml
# serverless.yml

type: docker-image-test

components:
  docker:
    type: docker-image
    inputs:
      dockerfilePath: ${self.path}/Dockerfile
      contextPath: ${self.path}
      tag: jdoe/my-image:latest
      registryUrl: https://my-registry.com # leave this empty if the image should be pushed to the public Docker registry
      username: jdoe
      password: s0m3p455w0rd
```

```
# Dockerfile

FROM hello-world
```

You can run `components build` / `components deploy` to build / push the Docker image. Once built and pushed you can remove the Docker image via `components remove`.

eahefnawy commented on registry/docker-image-assembly/serverless.yml:
```diff
@@ -0,0 +1,48 @@
+  registryUrl:
+    type: string
+    displayName: Registry URL
+    description: The Docker registry URL which should be used to push the image (leave empty if you want to push the image to the public Docker registry)
```
shouldn't there be a default here?

pmuens replied:
That's a good point 👍 I have added the default in the code, but putting it here makes a lot more sense. I'll update it...

eahefnawy commented on registry/docker-image-assembly/serverless.yml:
```diff
@@ -0,0 +1,48 @@
+    example: .
+  tag:
+    type: string
+    required: true
```
oh is a tag required?!

pmuens replied:
Yes, the tag must be set. Otherwise pushing it to the registry is not possible...

eahefnawy commented on registry/docker-image-assembly/src/utils/checkDockerSetup.test.js:
```diff
@@ -0,0 +1,46 @@
+  it('should throw if Docker is not installed', async () => {
+    isDockerInstalled.mockImplementation(() => false)
+    // NOTE: docker can't be running if it's not installed
+    // isDockerRunning.mockImplementation(() => false)
```
should this comment be removed?

pmuens replied:
Yes, we can remove it. I kept it as a companion code snippet for the `NOTE:` comment above.

pmuens commented:
@eahefnawy @raeesbhatti thanks for the review and the comments!

I've rebased the PR (since I merged another PR which introduced linting enhancements) and addressed all the review comments. Just tested this PR again and it works fine with the provided config in the PR description :+1:
